packages/marketo-adapter/src/adapter.ts

```diff
@@ -0,0 +1,180 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import {
+  FetchResult,
+  AdapterOperations,
+  ChangeGroup,
+  DeployResult,
+  InstanceElement,
+  Element, ObjectType, ElemID, BuiltinTypes, Values,
+} from '@salto-io/adapter-api'
+import { deployInstance, resolveValues } from '@salto-io/adapter-utils'
+import MarketoClient from './client/client'
+import { createMarketoObjectType, getLookUpName, Types } from './transformers/transformer'
+import { LeadAttribute, CustomObject, MarketoMetadata } from './client/types'
+import { MARKETO, OBJECTS_NAMES, TYPES_PATH } from './constants'
+
+
+export interface MarketoAdapterParams {
+  client: MarketoClient
+}
+
+export default class MarketoAdapter implements AdapterOperations {
+  private readonly client: MarketoClient
+
+  public constructor({
+    client,
+  }: MarketoAdapterParams) {
+    this.client = client
+  }
+
+  /**
+   * Fetch configuration elements: objects, types and instances for the given Marketo account.
+   */
+  public async fetch(): Promise<FetchResult> {
+    const fieldTypes = Types.getAllFieldTypes()
+    const objectTypes = await this.getAllMarketoObjects()
+
+    const elements = [...fieldTypes, ...objectTypes] as Element[]
+    return { elements }
+  }
+
+  private async getAllMarketoObjects(): Promise<ObjectType[]> {
+    return (await Promise.all(
+      Object.values(OBJECTS_NAMES)
+        .map(t => this.getAllMarketoObjectTypes(t))
+    )).flat()
+  }
+
+  private async getAllMarketoObjectTypes(type: string): Promise<ObjectType[]> {
+    switch (type) {
+      case OBJECTS_NAMES.CUSTOM_OBJECT:
+        return (await this.getAllCustomObjects())
+          .map((co: CustomObject) =>
+            createMarketoObjectType(co.name, co.fields))
+      case OBJECTS_NAMES.LEAD:
+        return [createMarketoObjectType(type,
+          await this.client.describe(type) as LeadAttribute[])]
+      default:
+        throw new Error(`Unsupported type: ${type}`)
+    }
+  }
+
+  private async getAllCustomObjects(): Promise<CustomObject[]> {
+    const customObjectsMetadata = await this.client
+      .getAllInstances(OBJECTS_NAMES.CUSTOM_OBJECT) as CustomObject[]
+    if (customObjectsMetadata === undefined) {
+      return []
+    }
+    return (await Promise.all(
+      customObjectsMetadata.map((co: CustomObject): Promise<Values[]> =>
+        this.client.describe(`${OBJECTS_NAMES.CUSTOM_OBJECT}/${co.name}`))
+    )).flat() as CustomObject[]
```

Comment:
style nit:
IMO we shouldn't declare variable's types when the compiler can deduce them by itself.
The compiler already 'knows' that `customObjectsMetadata` is ` CustomObject[]` and also knows that `this.client.describe()` returns `Promise<Values[]>`.
I think that this extra code doesn't really help reading the code and just makes it longer and less readable.
I think we should change it all over the PR to align with all of our code base but as I said it's just a style preference :)

```suggestion
    return (await Promise.all(
      customObjectsMetadata.map(co =>
        this.client.describe(`${OBJECTS_NAMES.CUSTOM_OBJECT}/${co.name}`))
    )).flat() as CustomObject[]
```